File: soln-graph-problems/all-paths-from-source-to-target-using-dfs.java
/*
https://practice.geeksforgeeks.org/problems/count-the-paths4332/1
https://leetcode.com/problems/all-paths-from-source-to-target/submissions/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    boolean[] visited;
    int[][] adj_list;
    List<List<Integer>> possible_paths=new ArrayList<>();
    
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        if(graph==null)
            return null;
        
        int s=0;
        int d=0;
        for(int edges[]:graph)
            for(int v:edges)
                d=Math.max(d,v);
        int n=d+1;        
        adj_list=graph;
        visited=new boolean[n];     
        List<Integer> path=new ArrayList<>();
        path.add(s);
        dfs(s,d,path);
        return possible_paths;
    }
    
    public void dfs(int source, int dest, List<Integer> path){
        if(source==dest){
            possible_paths.add(new ArrayList<>(path));            
            return;            
        }
  
        visited[source]=true;

        for(int adj_node:adj_list[source]){
            if(!visited[adj_node]){        
                path.add(adj_node);
                dfs(adj_node,dest,path);
                path.remove(path.size()-1);
            }
        }
        visited[source]=false;
    }
}

File: C++/soln-graph-problems/all-paths-from-source-to-target-using-dfs.cpp
/**
 @author Farheen Bano
 
 Date 24-07-2021
  
 References-
 https://practice.geeksforgeeks.org/problems/count-the-paths4332/1
 https://leetcode.com/problems/all-paths-from-source-to-target/submissions/
*/

class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int dest=graph.size()-1;
        int source=0;
        
        vector<vector<int>> result;
        vector<int> path;
        dfs(graph,result,path,source,dest);
        return result;
    }
    
    void dfs(vector<vector<int>> &graph, vector<vector<int>> &result,
             vector<int> &path, int u,int dest){
        path.push_back(u);
        if(u==dest)
            result.push_back(path);
        else
        {
            for(int v:graph[u]){
                dfs(graph,result,path,v,dest);
            }
        }
        
        path.pop_back();
        
    }
};